### ISSUE TITLE
[Bug]: Date Picker not usable for Selecting Birth Date, despite this use case being advertised.

### ISSUE THREAD

**Issue by johanneskares** [shadcn-ui/ui#2982]
### Describe the bug

Has anyone actually tried to use the date picker for selecting your bith date?

<img width="459" alt="Screenshot 2024-03-11 at 16 50 55" src="https://github.com/shadcn-ui/ui/assets/1416628/58e84d71-0dc2-42e2-bf73-4f5ccda0134a">

It's basically impossible, because you can't select the year.

### Affected component/components

Date Picker, Calendar

### How to reproduce

Obvious

### System Info

```bash
all browsers
```

### Before submitting

- [X] I've made research efforts and searched the documentation
- [X] I've searched for existing issues

**Comment by johanneskares** [comment]
For anyone looking for a solution, here is a small modification to the component that works for me.

<img width="305" alt="Screenshot 2024-03-11 at 17 43 22" src="https://github.com/shadcn-ui/ui/assets/1416628/8c1c3a37-b99c-423a-8286-bb8061fa3804">

```typescript
<Calendar
    selected={field.value}
    onSelect={field.onChange}
    disabled={(date) => date > new Date() || date < new Date("1900-01-01")}
    captionLayout="dropdown"
    toYear={2010}
    fromYear={1900}
    classNames={{
        day_hidden: "invisible",
        dropdown: "px-2 py-1.5 rounded-md bg-popover text-popover-foreground text-sm  focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 ring-offset-background",
        caption_dropdowns: "flex gap-3",
        vhidden: "hidden",
        caption_label: "hidden",
    }}
    initialFocus
/>
```

**Comment by TheZodo** [comment]
really nice @johanneskares I guess I took a hard route and just used another component. Tremendously helpful!

**Comment by Nikkakko** [comment]
> ```ts
>  captionLayout="dropdown"
>     toYear={2010}
>     fromYear={1900}
>     classNames={{
>         day_hidden: "invisible",
>         dropdown: "px-2 py-1.5 rounded-md bg-popover text-popover-foreground text-sm  focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 ring-offset-background",
>         caption_dropdowns: "flex gap-3",
>         vhidden: "hidden",
>         caption_label: "hidden",
>     }}
> ```

its fine but when you select calendar and re-open , its reseted. how can we fix it?

**Comment by camposvictor** [comment]
> @johanneskares its fine but when you select calendar and re-open , its reseted. how can we fix it?

https://react-day-picker.js.org/basics/navigation

you have to pass the current field value to the `defaultMonth` prop.

**Comment by nathanielrich** [comment]
it would be great when this works for range too... 
if i change the date in the caption it changes for both calendar views...

**Comment by dmr** [comment]
This is a great ticket and solution, thank you!

**Comment by DallasBelt** [comment]
Thanks! Hopefully the calendar can work as expected with the property `captionLayout='dropdown'` with an update.

**Comment by flixlix** [comment]
feel free to use this variant. you will need `react-day-picker` v9 or above:
https://github.com/shadcn-ui/ui/pull/4421

I think this is better UX than a Select Dropdown, it is inspired by: https://material.angular.io/components/datepicker/overview

**Comment by carlos-lopez-SaRa** [comment]
> For anyone looking for a solution, here is a small modification to the component that works for me.
> 
> <img alt="Screenshot 2024-03-11 at 17 43 22" width="305" src="https://private-user-images.githubusercontent.com/1416628/311788009-8c1c3a37-b99c-423a-8286-bb8061fa3804.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2MDQxNTcsIm5iZiI6MTcyODYwMzg1NywicGF0aCI6Ii8xNDE2NjI4LzMxMTc4ODAwOS04YzFjM2EzNy1iOTljLTQyM2EtODI4Ni1iYjgwNjFmYTM4MDQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTBUMjM0NDE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2M5MTJkMDFjZGQxNjkyOTIwNTY1MTY5NDhjOWUyZGEzOTM3NWJmZGUzM2M2OWUxMjFlOTAzNmRjZWJmMGE5OSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xC5OxuSDj5Kk373FhN8xQmU9nqlh9mHvO-maqoMXq94">
> ```ts
> <Calendar
>     selected={field.value}
>     onSelect={field.onChange}
>     disabled={(date) => date > new Date() || date < new Date("1900-01-01")}
>     captionLayout="dropdown"
>     toYear={2010}
>     fromYear={1900}
>     classNames={{
>         day_hidden: "invisible",
>         dropdown: "px-2 py-1.5 rounded-md bg-popover text-popover-foreground text-sm  focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 ring-offset-background",
>         caption_dropdowns: "flex gap-3",
>         vhidden: "hidden",
>         caption_label: "hidden",
>     }}
>     initialFocus
> />
> ```
> 
> I'm happy to create a PR to add this as a variant.

As a stopgap solution this works great, hope the new calendar is ready soon.

**Comment by huybuidac** [comment]
If you're looking for something fancier and more functional, give this a try.

https://shadcn-datetime-picker-xi.vercel.app/
https://github.com/huybuidac/shadcn-datetime-picker

https://github.com/user-attachments/assets/7441e1ef-f03f-42a0-ab52-023dfa4e0f75

**Comment by carlos-lopez-SaRa** [comment]
I recently came across a requirement, was asked to allow for the date to be typed as well, not just have a select calendar, I have yet to see that being mentioned in here...

**Comment by huybuidac** [comment]
> I recently came across a requirement, was asked to allow for the date to be typed with keyboard as well, not just have a select calendar, I have yet to see that being mentioned in here...

first version

https://shadcn-datetime-picker-pro.vercel.app/?path=/story/datetime-picker--date-time-input-picker

https://github.com/user-attachments/assets/aa081186-a6cb-4d15-ac44-2aac3b2e052e

**Comment by carlos-lopez-SaRa** [comment]
Have you tried using it inside a form? I did a custom input as well but one of the issues I found latter was if I fill the input, then delete a couple of letters, I ended up having 1 of 2 problems (depending on how I handled the change event):

- Since Date objects have no in between, it's either a valid Date or `undefined`, on submit the default values of `react-use-form` would take over and reset the input to empty date (if the user wanted to correct 1 segment, the user would have to start typing from scratch).
- On submit, instead of the above, the input would reset to the previous valid value, unless the user deleted all characters of the input.

Basically in my case the form's control value and the input would be out of sync, and ended up having to transform the date into a date string.

**Comment by huybuidac** [comment]
> @huybuidac Does that include support for only date as well? what about only time input?

you can use format = "hh:mm:ss"

**Comment by carlos-lopez-SaRa** [comment]
> > @huybuidac Does that include support for only date as well? what about only time input?
> 
> you can use format = "hh:mm:ss"

What would be the output data type? I doubt it would also be a date (assuming a time zone unaware time).

**Comment by huybuidac** [comment]
> > > @huybuidac Does that include support for only date as well? what about only time input?
> > 
> > 
> > you can use format = "hh:mm:ss"
> 
> What would be the output data type? I doubt it would also be a Date Object (assuming a time zone unaware time).

any timezone you want

**Comment by huybuidac** [comment]
> Have you tried using it inside a form? I did a custom input as well but one of the issues I found latter was if I fill the input, then delete a couple of letters, I ended up having 1 of 2 problems (depending on how I handled the change event):
> 
> * Since Date objects have no in between, it's either a valid Date or `undefined`, on submit the default values of `react-use-form` would take over and reset the input to empty date (if the user wanted to correct 1 segment, the user would have to start typing from scratch).
> * On submit, instead of the above, the input would reset to the previous valid value, unless the user deleted all characters of the input.
> 
> Basically in my case the form's control value and the input would be out of sync, and ended up having to transform the date into a date string.

may work https://shadcn-datetime-picker-pro.vercel.app/?path=/story/datetimepicker--date-time-input-picker-in-form

If you have any issues or ideas, create them in my repo. I'll handle them if I can.

**Comment by carlos-lopez-SaRa** [comment]
> > > > @huybuidac Does that include support for only date as well? what about only time input?
> > > 
> > > 
> > > you can use format = "hh:mm:ss"
> > 
> > 
> > What would be the output data type? I doubt it would also be a Date Object (assuming a time zone unaware time).
> 
> any timezone you want

What I'm saying is a whole Date object for a time input feels like a bit much, what if you only want the time part?

**Comment by huybuidac** [comment]
> What I'm saying is a whole Date object for a time input feels like a bit much, what if you only want the time part?

Like some other libraries, your logic should handle hh:mm:ss a format.
Alternatively, you can store the string value as hh:mm on the server, then merge it into a date object on the FE.

https://ant.design/components/time-picker

**Comment by shadcn** [comment]
This issue has been automatically closed due to one year of inactivity. If you’re still experiencing a similar problem or have additional details to share, please open a new issue following our current issue template. Your updated report helps us investigate and address concerns more efficiently. Thank you for your understanding! (This is an automated message)
